# Comparing `tmp/igs_toolbox-1.5.2-py3-none-any.whl.zip` & `tmp/igs_toolbox-1.5.7-py3-none-any.whl.zip`

## zipinfo {}

```diff
@@ -1,46 +1,86 @@
-Zip file size: 405738 bytes, number of entries: 44
--rw-r--r--  2.0 unx       78 b- defN 23-Dec-01 14:45 igs_toolbox/__init__.py
--rw-r--r--  2.0 unx        0 b- defN 23-Dec-01 14:45 igs_toolbox/py.typed
--rw-r--r--  2.0 unx        0 b- defN 23-Dec-01 14:45 igs_toolbox/converters/__init__.py
--rw-r--r--  2.0 unx     2460 b- defN 23-Dec-01 14:45 igs_toolbox/converters/convert_answersets.py
--rw-r--r--  2.0 unx     6128 b- defN 23-Dec-01 14:45 igs_toolbox/converters/convert_seq_metadata.py
--rw-r--r--  2.0 unx        0 b- defN 23-Dec-01 14:45 igs_toolbox/extractor/__init__.py
--rw-r--r--  2.0 unx     1783 b- defN 23-Dec-01 14:45 igs_toolbox/extractor/read_qr.py
--rw-r--r--  2.0 unx     5600 b- defN 23-Dec-01 14:45 igs_toolbox/extractor/res/extractor_sample_a.pdf
--rw-r--r--  2.0 unx     3956 b- defN 23-Dec-01 14:45 igs_toolbox/extractor/res/extractor_sample_b.pdf
--rw-r--r--  2.0 unx   508941 b- defN 23-Dec-01 14:45 igs_toolbox/extractor/res/extractor_sample_c.pdf
--rw-r--r--  2.0 unx        0 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/__init__.py
--rw-r--r--  2.0 unx     3549 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/json_checker.py
--rw-r--r--  2.0 unx    33480 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/seq_metadata_schema.py
--rw-r--r--  2.0 unx     9139 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/json/answerSetACBP.json
--rw-r--r--  2.0 unx     3588 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/json/answerSetCVDP.json
--rw-r--r--  2.0 unx    34242 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/json/answerSetEBCP.json
--rw-r--r--  2.0 unx     8959 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/json/answerSetEHCP.json
--rw-r--r--  2.0 unx    10818 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/json/answerSetLEGP.json
--rw-r--r--  2.0 unx     2733 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/json/answerSetLISP.json
--rw-r--r--  2.0 unx     3509 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/json/answerSetMSVP.json
--rw-r--r--  2.0 unx     2845 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/json/answerSetMYTP.json
--rw-r--r--  2.0 unx     2359 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/json/answerSetNEIP.json
--rw-r--r--  2.0 unx   322908 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/json/answerSetSALP.json
--rw-r--r--  2.0 unx    15030 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/json/answerSetSPNP.json
--rw-r--r--  2.0 unx      833 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/json/answerSetSTYP.json
--rw-r--r--  2.0 unx     2792 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/txt/answerSetACBP.txt
--rw-r--r--  2.0 unx       35 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/txt/answerSetCDFP.txt
--rw-r--r--  2.0 unx      927 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/txt/answerSetCVDP.txt
--rw-r--r--  2.0 unx     9497 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/txt/answerSetEBCP.txt
--rw-r--r--  2.0 unx     3767 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/txt/answerSetEHCP.txt
--rw-r--r--  2.0 unx     3211 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/txt/answerSetLEGP.txt
--rw-r--r--  2.0 unx      685 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/txt/answerSetLISP.txt
--rw-r--r--  2.0 unx      946 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/txt/answerSetMSVP.txt
--rw-r--r--  2.0 unx      643 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/txt/answerSetMYTP.txt
--rw-r--r--  2.0 unx      568 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/txt/answerSetNEIP.txt
--rw-r--r--  2.0 unx    95335 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/txt/answerSetSALP.txt
--rw-r--r--  2.0 unx     5721 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/txt/answerSetSPNP.txt
--rw-r--r--  2.0 unx       84 b- defN 23-Dec-01 14:45 igs_toolbox/formatChecker/res/species/txt/answerSetSTYP.txt
--rw-r--r--  2.0 unx     1068 b- defN 23-Dec-01 14:46 igs_toolbox-1.5.2.dist-info/LICENSE
--rw-r--r--  2.0 unx     4346 b- defN 23-Dec-01 14:46 igs_toolbox-1.5.2.dist-info/METADATA
--rw-r--r--  2.0 unx       92 b- defN 23-Dec-01 14:46 igs_toolbox-1.5.2.dist-info/WHEEL
--rw-r--r--  2.0 unx      190 b- defN 23-Dec-01 14:46 igs_toolbox-1.5.2.dist-info/entry_points.txt
--rw-r--r--  2.0 unx       12 b- defN 23-Dec-01 14:46 igs_toolbox-1.5.2.dist-info/top_level.txt
--rw-rw-r--  2.0 unx     4658 b- defN 23-Dec-01 14:46 igs_toolbox-1.5.2.dist-info/RECORD
-44 files, 1117515 bytes uncompressed, 397924 bytes compressed:  64.4%
+Zip file size: 440982 bytes, number of entries: 84
+-rw-r--r--  2.0 unx       78 b- defN 24-Apr-10 14:30 igs_toolbox/__init__.py
+-rw-r--r--  2.0 unx      405 b- defN 24-Apr-10 14:30 igs_toolbox/file_utils.py
+-rw-r--r--  2.0 unx      400 b- defN 24-Apr-10 14:30 igs_toolbox/log_utils.py
+-rw-r--r--  2.0 unx        0 b- defN 24-Apr-10 14:30 igs_toolbox/py.typed
+-rw-r--r--  2.0 unx      245 b- defN 24-Apr-10 14:30 igs_toolbox/version_utils.py
+-rw-r--r--  2.0 unx        0 b- defN 24-Apr-10 14:30 igs_toolbox/converters/__init__.py
+-rw-r--r--  2.0 unx     6463 b- defN 24-Apr-10 14:30 igs_toolbox/converters/convert_seq_metadata.py
+-rw-r--r--  2.0 unx     3574 b- defN 24-Apr-10 14:30 igs_toolbox/converters/convert_valuesets.py
+-rw-r--r--  2.0 unx        0 b- defN 24-Apr-10 14:30 igs_toolbox/extractor/__init__.py
+-rw-r--r--  2.0 unx     2205 b- defN 24-Apr-10 14:30 igs_toolbox/extractor/read_qr.py
+-rw-r--r--  2.0 unx     5600 b- defN 24-Apr-10 14:30 igs_toolbox/extractor/res/extractor_sample_a.pdf
+-rw-r--r--  2.0 unx     3956 b- defN 24-Apr-10 14:30 igs_toolbox/extractor/res/extractor_sample_b.pdf
+-rw-r--r--  2.0 unx   508941 b- defN 24-Apr-10 14:30 igs_toolbox/extractor/res/extractor_sample_c.pdf
+-rw-r--r--  2.0 unx        0 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/__init__.py
+-rw-r--r--  2.0 unx     5746 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/json_checker.py
+-rw-r--r--  2.0 unx    13515 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/seq_metadata_schema.py
+-rw-r--r--  2.0 unx    43776 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/json/materialACBP.json
+-rw-r--r--  2.0 unx    30845 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/json/materialCVDP.json
+-rw-r--r--  2.0 unx    43022 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/json/materialEBCP.json
+-rw-r--r--  2.0 unx     3676 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/json/materialEHCP.json
+-rw-r--r--  2.0 unx    12871 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/json/materialHAVP.json
+-rw-r--r--  2.0 unx     6019 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/json/materialHEVP.json
+-rw-r--r--  2.0 unx    27836 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/json/materialINVP.json
+-rw-r--r--  2.0 unx    14495 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/json/materialLEGP.json
+-rw-r--r--  2.0 unx    10868 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/json/materialLISP.json
+-rw-r--r--  2.0 unx     7667 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/json/materialMRAP.json
+-rw-r--r--  2.0 unx    11259 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/json/materialMSVP.json
+-rw-r--r--  2.0 unx    45206 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/json/materialMYTP.json
+-rw-r--r--  2.0 unx    12690 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/json/materialNEIP.json
+-rw-r--r--  2.0 unx     7414 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/json/materialSALP.json
+-rw-r--r--  2.0 unx     7154 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/json/materialSPNP.json
+-rw-r--r--  2.0 unx    11308 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/json/materialSTYP.json
+-rw-r--r--  2.0 unx     4387 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/json/materialWNVP.json
+-rw-r--r--  2.0 unx     2017 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetACBP.txt
+-rw-r--r--  2.0 unx       25 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetCDVP.txt
+-rw-r--r--  2.0 unx     1633 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetCVDP.txt
+-rw-r--r--  2.0 unx     1991 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetEBCP.txt
+-rw-r--r--  2.0 unx      110 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetEHCP.txt
+-rw-r--r--  2.0 unx      529 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetHAVP.txt
+-rw-r--r--  2.0 unx      244 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetHEVP.txt
+-rw-r--r--  2.0 unx      433 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetHIVP.txt
+-rw-r--r--  2.0 unx     1253 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetINVP.txt
+-rw-r--r--  2.0 unx      644 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetLEGP.txt
+-rw-r--r--  2.0 unx      418 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetLISP.txt
+-rw-r--r--  2.0 unx      278 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetMRAP.txt
+-rw-r--r--  2.0 unx      413 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetMSVP.txt
+-rw-r--r--  2.0 unx     2070 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetMYTP.txt
+-rw-r--r--  2.0 unx      473 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetNEGP.txt
+-rw-r--r--  2.0 unx      628 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetNEIP.txt
+-rw-r--r--  2.0 unx      263 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetSALP.txt
+-rw-r--r--  2.0 unx      318 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetSPNP.txt
+-rw-r--r--  2.0 unx      422 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetSTYP.txt
+-rw-r--r--  2.0 unx      144 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/isolation_source/txt/valueSetWNVP.txt
+-rw-r--r--  2.0 unx     9139 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/json/answerSetACBP.json
+-rw-r--r--  2.0 unx     3588 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/json/answerSetCVDP.json
+-rw-r--r--  2.0 unx    34242 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/json/answerSetEBCP.json
+-rw-r--r--  2.0 unx     8959 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/json/answerSetEHCP.json
+-rw-r--r--  2.0 unx    10818 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/json/answerSetLEGP.json
+-rw-r--r--  2.0 unx     2733 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/json/answerSetLISP.json
+-rw-r--r--  2.0 unx     3509 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/json/answerSetMSVP.json
+-rw-r--r--  2.0 unx     2845 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/json/answerSetMYTP.json
+-rw-r--r--  2.0 unx     2359 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/json/answerSetNEIP.json
+-rw-r--r--  2.0 unx   322908 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/json/answerSetSALP.json
+-rw-r--r--  2.0 unx    15030 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/json/answerSetSPNP.json
+-rw-r--r--  2.0 unx      833 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/json/answerSetSTYP.json
+-rw-r--r--  2.0 unx     2792 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/txt/valueSetACBP.txt
+-rw-r--r--  2.0 unx       35 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/txt/valueSetCDFP.txt
+-rw-r--r--  2.0 unx      927 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/txt/valueSetCVDP.txt
+-rw-r--r--  2.0 unx     9497 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/txt/valueSetEBCP.txt
+-rw-r--r--  2.0 unx     3767 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/txt/valueSetEHCP.txt
+-rw-r--r--  2.0 unx     3211 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/txt/valueSetLEGP.txt
+-rw-r--r--  2.0 unx      685 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/txt/valueSetLISP.txt
+-rw-r--r--  2.0 unx      946 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/txt/valueSetMSVP.txt
+-rw-r--r--  2.0 unx      643 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/txt/valueSetMYTP.txt
+-rw-r--r--  2.0 unx      568 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/txt/valueSetNEIP.txt
+-rw-r--r--  2.0 unx    95335 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/txt/valueSetSALP.txt
+-rw-r--r--  2.0 unx     5721 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/txt/valueSetSPNP.txt
+-rw-r--r--  2.0 unx       84 b- defN 24-Apr-10 14:30 igs_toolbox/formatChecker/res/species/txt/valueSetSTYP.txt
+-rw-r--r--  2.0 unx     1068 b- defN 24-Apr-10 14:32 igs_toolbox-1.5.7.dist-info/LICENSE
+-rw-r--r--  2.0 unx     4440 b- defN 24-Apr-10 14:32 igs_toolbox-1.5.7.dist-info/METADATA
+-rw-r--r--  2.0 unx       92 b- defN 24-Apr-10 14:32 igs_toolbox-1.5.7.dist-info/WHEEL
+-rw-r--r--  2.0 unx      255 b- defN 24-Apr-10 14:32 igs_toolbox-1.5.7.dist-info/entry_points.txt
+-rw-r--r--  2.0 unx       12 b- defN 24-Apr-10 14:32 igs_toolbox-1.5.7.dist-info/top_level.txt
+-rw-rw-r--  2.0 unx     9506 b- defN 24-Apr-10 14:32 igs_toolbox-1.5.7.dist-info/RECORD
+84 files, 1422474 bytes uncompressed, 424976 bytes compressed:  70.1%
```

## zipnote {}

```diff
@@ -1,22 +1,31 @@
 Filename: igs_toolbox/__init__.py
 Comment: 
 
+Filename: igs_toolbox/file_utils.py
+Comment: 
+
+Filename: igs_toolbox/log_utils.py
+Comment: 
+
 Filename: igs_toolbox/py.typed
 Comment: 
 
-Filename: igs_toolbox/converters/__init__.py
+Filename: igs_toolbox/version_utils.py
 Comment: 
 
-Filename: igs_toolbox/converters/convert_answersets.py
+Filename: igs_toolbox/converters/__init__.py
 Comment: 
 
 Filename: igs_toolbox/converters/convert_seq_metadata.py
 Comment: 
 
+Filename: igs_toolbox/converters/convert_valuesets.py
+Comment: 
+
 Filename: igs_toolbox/extractor/__init__.py
 Comment: 
 
 Filename: igs_toolbox/extractor/read_qr.py
 Comment: 
 
 Filename: igs_toolbox/extractor/res/extractor_sample_a.pdf
@@ -33,14 +42,125 @@
 
 Filename: igs_toolbox/formatChecker/json_checker.py
 Comment: 
 
 Filename: igs_toolbox/formatChecker/seq_metadata_schema.py
 Comment: 
 
+Filename: igs_toolbox/formatChecker/res/isolation_source/json/materialACBP.json
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/json/materialCVDP.json
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/json/materialEBCP.json
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/json/materialEHCP.json
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/json/materialHAVP.json
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/json/materialHEVP.json
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/json/materialINVP.json
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/json/materialLEGP.json
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/json/materialLISP.json
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/json/materialMRAP.json
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/json/materialMSVP.json
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/json/materialMYTP.json
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/json/materialNEIP.json
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/json/materialSALP.json
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/json/materialSPNP.json
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/json/materialSTYP.json
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/json/materialWNVP.json
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetACBP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetCDVP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetCVDP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetEBCP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetEHCP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetHAVP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetHEVP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetHIVP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetINVP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetLEGP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetLISP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetMRAP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetMSVP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetMYTP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetNEGP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetNEIP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetSALP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetSPNP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetSTYP.txt
+Comment: 
+
+Filename: igs_toolbox/formatChecker/res/isolation_source/txt/valueSetWNVP.txt
+Comment: 
+
 Filename: igs_toolbox/formatChecker/res/species/json/answerSetACBP.json
 Comment: 
 
 Filename: igs_toolbox/formatChecker/res/species/json/answerSetCVDP.json
 Comment: 
 
 Filename: igs_toolbox/formatChecker/res/species/json/answerSetEBCP.json
@@ -69,65 +189,65 @@
 
 Filename: igs_toolbox/formatChecker/res/species/json/answerSetSPNP.json
 Comment: 
 
 Filename: igs_toolbox/formatChecker/res/species/json/answerSetSTYP.json
 Comment: 
 
-Filename: igs_toolbox/formatChecker/res/species/txt/answerSetACBP.txt
+Filename: igs_toolbox/formatChecker/res/species/txt/valueSetACBP.txt
 Comment: 
 
-Filename: igs_toolbox/formatChecker/res/species/txt/answerSetCDFP.txt
+Filename: igs_toolbox/formatChecker/res/species/txt/valueSetCDFP.txt
 Comment: 
 
-Filename: igs_toolbox/formatChecker/res/species/txt/answerSetCVDP.txt
+Filename: igs_toolbox/formatChecker/res/species/txt/valueSetCVDP.txt
 Comment: 
 
-Filename: igs_toolbox/formatChecker/res/species/txt/answerSetEBCP.txt
+Filename: igs_toolbox/formatChecker/res/species/txt/valueSetEBCP.txt
 Comment: 
 
-Filename: igs_toolbox/formatChecker/res/species/txt/answerSetEHCP.txt
+Filename: igs_toolbox/formatChecker/res/species/txt/valueSetEHCP.txt
 Comment: 
 
-Filename: igs_toolbox/formatChecker/res/species/txt/answerSetLEGP.txt
+Filename: igs_toolbox/formatChecker/res/species/txt/valueSetLEGP.txt
 Comment: 
 
-Filename: igs_toolbox/formatChecker/res/species/txt/answerSetLISP.txt
+Filename: igs_toolbox/formatChecker/res/species/txt/valueSetLISP.txt
 Comment: 
 
-Filename: igs_toolbox/formatChecker/res/species/txt/answerSetMSVP.txt
+Filename: igs_toolbox/formatChecker/res/species/txt/valueSetMSVP.txt
 Comment: 
 
-Filename: igs_toolbox/formatChecker/res/species/txt/answerSetMYTP.txt
+Filename: igs_toolbox/formatChecker/res/species/txt/valueSetMYTP.txt
 Comment: 
 
-Filename: igs_toolbox/formatChecker/res/species/txt/answerSetNEIP.txt
+Filename: igs_toolbox/formatChecker/res/species/txt/valueSetNEIP.txt
 Comment: 
 
-Filename: igs_toolbox/formatChecker/res/species/txt/answerSetSALP.txt
+Filename: igs_toolbox/formatChecker/res/species/txt/valueSetSALP.txt
 Comment: 
 
-Filename: igs_toolbox/formatChecker/res/species/txt/answerSetSPNP.txt
+Filename: igs_toolbox/formatChecker/res/species/txt/valueSetSPNP.txt
 Comment: 
 
-Filename: igs_toolbox/formatChecker/res/species/txt/answerSetSTYP.txt
+Filename: igs_toolbox/formatChecker/res/species/txt/valueSetSTYP.txt
 Comment: 
 
-Filename: igs_toolbox-1.5.2.dist-info/LICENSE
+Filename: igs_toolbox-1.5.7.dist-info/LICENSE
 Comment: 
 
-Filename: igs_toolbox-1.5.2.dist-info/METADATA
+Filename: igs_toolbox-1.5.7.dist-info/METADATA
 Comment: 
 
-Filename: igs_toolbox-1.5.2.dist-info/WHEEL
+Filename: igs_toolbox-1.5.7.dist-info/WHEEL
 Comment: 
 
-Filename: igs_toolbox-1.5.2.dist-info/entry_points.txt
+Filename: igs_toolbox-1.5.7.dist-info/entry_points.txt
 Comment: 
 
-Filename: igs_toolbox-1.5.2.dist-info/top_level.txt
+Filename: igs_toolbox-1.5.7.dist-info/top_level.txt
 Comment: 
 
-Filename: igs_toolbox-1.5.2.dist-info/RECORD
+Filename: igs_toolbox-1.5.7.dist-info/RECORD
 Comment: 
 
 Zip file comment:
```

## igs_toolbox/converters/convert_seq_metadata.py

```diff
@@ -1,78 +1,79 @@
 import json
 import logging
 import sys
+import warnings
 from datetime import datetime
 from pathlib import Path
-import warnings
+from typing import Any, Dict, List, Optional
+
+from igs_toolbox.log_utils import setup_logging
 
 if sys.version_info >= (3, 10):
-    from typing import Annotated, Dict, Optional
+    from typing import Annotated
 else:
-    from typing_extensions import Annotated, Dict, Optional
+    from typing_extensions import Annotated
 
 import pandas as pd
 import typer
 
 if sys.version_info >= (3, 9):
     from zoneinfo import ZoneInfo
 else:
     from backports.zoneinfo import ZoneInfo
 
-import igs_toolbox
+
 from igs_toolbox.formatChecker import json_checker
 from igs_toolbox.formatChecker.seq_metadata_schema import SeqMetadataKeys, ValidationError
+from igs_toolbox.version_utils import version_callback
 
-sys.path.append(Path(__file__).resolve().parent.parent)
-
-NOW = datetime.now(tz=ZoneInfo("Europe/Berlin")).strftime("%Y-%m-%dT%H-%M-%S")
-
-
-def version_callback(value: bool) -> None:  # noqa: FBT001
-    """Print toolbox version."""
-    if value:
-        print(f"IGS Toolbox Version: {igs_toolbox.__version__}")  # noqa: T201
-        raise typer.Exit
+TIMESTAMP = datetime.now(tz=ZoneInfo("Europe/Berlin")).strftime("%Y-%m-%dT%H-%M-%S")  # type: ignore[abstract]
 
 
 app = typer.Typer()
 
 
-def nest_files_and_upload_entries(entry_dict: Dict[str, str]) -> None:
+def nest_files_and_upload_entries(entry_dict: Dict[str, str]) -> Dict[str, Any]:
     """Move files and uploads entries into list of dicts."""
     # If any file entry exists, add to dict
+    entry_dict_out: Dict[str, Any] = entry_dict.copy()
     if any(
         filename in entry_dict for filename in ["FILE_1_NAME", "FILE_1_SHA256SUM", "FILE_2_NAME", "FILE_2_SHA256SUM"]
     ):
-        entry_dict[SeqMetadataKeys.FILES.value] = []
+        entry_dict_out[SeqMetadataKeys.FILES] = []
         for idx in ["1", "2"]:
             file_info = {}
             for field in ["NAME", "SHA256SUM"]:
                 key = f"FILE_{idx}_{field}"
-                if key in entry_dict:
-                    file_info.update({f"FILE_{field}": entry_dict[key]})
-                    del entry_dict[key]
+                if key in entry_dict_out:
+                    file_info.update({f"FILE_{field}": entry_dict_out[key]})
+                    del entry_dict_out[key]
             if len(file_info) > 0:
-                entry_dict[SeqMetadataKeys.FILES.value].append(file_info)
+                entry_dict_out[SeqMetadataKeys.FILES].append(file_info)
 
     # If any upload entry exists, add to dict
     upload_keys = [
-        SeqMetadataKeys.UPLOAD_DATE.value,
-        SeqMetadataKeys.UPLOAD_STATUS.value,
-        SeqMetadataKeys.UPLOAD_SUBMITTER.value,
-        SeqMetadataKeys.REPOSITORY_ID.value,
-        SeqMetadataKeys.REPOSITORY_NAME.value,
-        SeqMetadataKeys.REPOSITORY_LINK.value,
+        SeqMetadataKeys.UPLOAD_DATE,
+        SeqMetadataKeys.UPLOAD_STATUS,
+        SeqMetadataKeys.UPLOAD_SUBMITTER,
+        SeqMetadataKeys.REPOSITORY_ID,
+        SeqMetadataKeys.REPOSITORY_NAME,
+        SeqMetadataKeys.REPOSITORY_LINK,
     ]
-    if any(filename in entry_dict for filename in upload_keys):
-        entry_dict[SeqMetadataKeys.UPLOADS.value] = [{}]
+    if any(fieldname in entry_dict_out for fieldname in upload_keys):
+        entry_dict_out[SeqMetadataKeys.UPLOADS] = [{}]
         for field in upload_keys:
-            if field in entry_dict:
-                entry_dict[SeqMetadataKeys.UPLOADS.value][0].update({field: entry_dict[field]})
-                del entry_dict[field]
+            if field in entry_dict_out:
+                entry_dict_out[SeqMetadataKeys.UPLOADS][0].update({field: entry_dict_out[field]})
+                del entry_dict_out[field]
+    return entry_dict_out
+
+
+def is_empty_string(value: Any) -> bool:  # noqa: ANN401
+    return isinstance(value, str) and not value.strip()
 
 
 @app.command(name="convertSeqMetadata", help="Convert table of seq metadata to json files.")
 def convert(
     input_file: Annotated[
         Path,
         typer.Option(
@@ -95,78 +96,89 @@
             file_okay=False,
             help="Path to output folder for json files.",
         ),
     ],
     log_file: Annotated[
         Path,
         typer.Option("--log_file", "-l", dir_okay=False, help="Path to log file."),
-    ] = Path(f"./convertSeqMetadata_{NOW}.log"),
+    ] = Path(f"./convertSeqMetadata_{TIMESTAMP}.log"),
     version: Annotated[  # noqa: ARG001
         Optional[bool],
-        typer.Option("--version", callback=version_callback, is_eager=True),
+        typer.Option("--version", "-V", callback=version_callback, is_eager=True),
     ] = None,
 ) -> None:
-    log_file.parent.mkdir(exist_ok=True, parents=True)
-    logging.basicConfig(
-        filename=log_file,
-        level=logging.ERROR,
-        format="%(asctime)s %(levelname)s %(message)s",
-        force=True,
-    )
+    setup_logging(log_file=log_file, debug=False)
 
+    # read table file
     meta_df: pd.DataFrame
     suffix = input_file.suffix
     if suffix == ".csv":
         meta_df = pd.read_csv(input_file, sep=",", dtype=str)
     elif suffix == ".tsv":
         meta_df = pd.read_csv(input_file, sep="\t", dtype=str)
     elif suffix == ".xlsx":
         with warnings.catch_warnings():
-            warnings.filterwarnings("ignore", category=UserWarning, message="Data Validation extension is not supported and will be removed")
+            warnings.filterwarnings(
+                "ignore",
+                category=UserWarning,
+                message="Data Validation extension is not supported and will be removed",
+            )
             meta_df = pd.read_excel(
                 input_file,
                 dtype=str,  # Cannot use parse dates here because missing optional columns will lead to an error.
             )
-
         for date_col in [
-            SeqMetadataKeys.DATE_OF_RECEIVING.value,
-            SeqMetadataKeys.DATE_OF_SAMPLING.value,
-            SeqMetadataKeys.DATE_OF_SEQUENCING.value,
-            SeqMetadataKeys.DATE_OF_SUBMISSION.value,
-            SeqMetadataKeys.UPLOAD_DATE.value,
+            SeqMetadataKeys.DATE_OF_RECEIVING,
+            SeqMetadataKeys.DATE_OF_SAMPLING,
+            SeqMetadataKeys.DATE_OF_SEQUENCING,
+            SeqMetadataKeys.DATE_OF_SUBMISSION,
+            SeqMetadataKeys.UPLOAD_DATE,
         ]:
             # Remove the time from the datetime
             if date_col in meta_df:
                 meta_df[date_col] = meta_df[date_col].str.replace(" 00:00:00", "")
     else:
         logging.error(f"Files of type {suffix} cannot be converted yet. Please provide either a xlsx, csv or tsv file.")
         raise typer.Abort
 
-    # Create output directory
-    output.mkdir(parents=True, exist_ok=True)
     # Convert to json
-    meta_dict: list[dict[str, str]] = meta_df.to_dict(orient="records")
+    metadata_dicts = []
+    raw_rows: List[Dict[str, str]] = meta_df.to_dict(orient="records")  # type: ignore[assignment]
+    for row_dict in raw_rows:
+        # remove empty strings and NANs
+        clean_dict = {key: value for key, value in row_dict.items() if not is_empty_string(value)}
+        clean_dict = {key: value for key, value in clean_dict.items() if not pd.isna(value)}
+
+        # transform file and upload entries into nested list
+        clean_dict = nest_files_and_upload_entries(clean_dict)
+
+        metadata_dicts.append(clean_dict)
+
+    # validate and write
+    output.mkdir(parents=True, exist_ok=True)
     failed = []
-    for idx, entry_dict in enumerate(meta_dict, start=1):
-        sample_id = entry_dict[SeqMetadataKeys.LAB_SEQUENCE_ID.value]
-        # replace NANs
-        clean_dict = {k: entry_dict[k] for k in entry_dict if not pd.isna(entry_dict[k])}
-        # move files and uploads into nested list
-        nest_files_and_upload_entries(clean_dict)
+    for idx, metadata_dict in enumerate(metadata_dicts, start=1):
+        # validate metadata
         try:
-            json_checker.check_seq_metadata(clean_dict)
-            with (output / f"{sample_id}_sequencing_metadata.json").open("w") as outfile:
-                json.dump(clean_dict, outfile, indent=4)
-        except ValidationError as err:
+            json_checker.check_seq_metadata(metadata_dict)
+        except ValidationError as e:
             failed.append(idx)
-            logging.error(f"Invalid data in row {idx}: {err}")
+            logging.error(f"Invalid data in row {idx}: {e}")  # noqa: TRY400
+            continue
+
+        # write metadata
+        sample_id = metadata_dict[SeqMetadataKeys.LAB_SEQUENCE_ID]
+        output_file = output / f"{sample_id}_sequencing_metadata.json"
+        output_file.write_text(json.dumps(metadata_dict, indent=4))
+
     if len(failed):
         logging.error(f"The following rows did not pass validation and were hence not converted: {failed}")
-        print(
-            f"Some rows did not pass validation, please consult the log file at {log_file.resolve()} for more information."
+        print(  # noqa: T201
+            f"Some rows did not pass validation, "
+            f"please consult the log file at {log_file.resolve()} for more information.",
         )
         raise typer.Exit
 
 
 def main() -> None:
     """Entry point of CLI tool."""
     app()
```

## igs_toolbox/extractor/read_qr.py

```diff
@@ -1,38 +1,65 @@
 # Requirements
 # mamba install poppler
 
 # Import modules
-import argparse
+
 import logging
 import sys
+from datetime import datetime
 from pathlib import Path
+from typing import List, Optional
 
 import cv2
-from pdf2image import convert_from_path
+import typer
+from pdf2image.pdf2image import convert_from_path
+
+from igs_toolbox.version_utils import version_callback
+
+if sys.version_info >= (3, 10):
+    from typing import Annotated
+
+    from zoneinfo import ZoneInfo
+else:
+    from backports.zoneinfo import ZoneInfo
+    from typing_extensions import Annotated
 
-import igs_toolbox
+NOW = datetime.now(tz=ZoneInfo("Europe/Berlin")).strftime("%Y-%m-%dT%H-%M-%S")  # type: ignore[abstract]
+app = typer.Typer()
 
 
 def read_qr_code(filename: str) -> str:
     try:
         img = cv2.imread(filename)
         detect = cv2.QRCodeDetector()
-        value, points, straight_qrcode = detect.detectAndDecode(img)
+        value, _, _ = detect.detectAndDecode(img)
     except Exception:  # noqa: BLE001
         return "Error during QR code detection."
     return value
 
 
-def main(args=None) -> None:  # noqa: ANN001
-    input_files = parse(args).files
+@app.command(name="readQR", help="Extract QR codes from files.")
+def read_qr_codes(
+    input_files: Annotated[
+        List[Path],
+        typer.Argument(
+            dir_okay=False,
+            file_okay=True,
+            help="Paths to input files.",
+        ),
+    ],
+    version: Annotated[  # noqa: ARG001
+        Optional[bool],
+        typer.Option("--version", "-V", callback=version_callback, is_eager=True),
+    ] = None,
+) -> None:
     for file in input_files:
-        if not Path(file).is_file():
+        if not file.is_file():
             logging.error(f"{file} does not point to a file. Aborting.")
-            sys.exit(1)
+            raise typer.Abort(1)
 
     # Iterate over files
     if len(input_files) > 0:
         for file in input_files:
             filename = Path(file).name.split(".")[0]
             images = convert_from_path(file)
 
@@ -42,25 +69,16 @@
                 images[i].save(page_name, "PNG")
 
                 # Detect QR code and print it
                 id_value = read_qr_code(page_name)
                 print(f"{filename}\t{id_value}")  # noqa: T201
                 Path(page_name).unlink()
     else:
-        sys.exit(2)
+        raise typer.Abort(2)
 
 
-# Read command line arguments
-def parse(args=None) -> argparse.Namespace:  # noqa: ANN001
-    parser = argparse.ArgumentParser(prog=Path(__file__).name.split(".")[0])
-    parser.add_argument("files", nargs="*")
-    parser.add_argument(
-        "-V",
-        "--version",
-        action="version",
-        version=f"%(prog)s {igs_toolbox.__version__}",
-    )
-    return parser.parse_args(args)
+def main() -> None:
+    app()
 
 
 if __name__ == "__main__":
     main()
```

## igs_toolbox/formatChecker/json_checker.py

```diff
@@ -1,111 +1,164 @@
-import argparse
 import json
 import logging
-import re
 import sys
 from pathlib import Path
-from typing import Any, Dict, Union
+from typing import Any, Dict, Optional
+
+import typer
+
+if sys.version_info >= (3, 10):
+    from typing import Annotated
+else:
+    from typing_extensions import Annotated
+from datetime import datetime
 
 import jsonschema
 
-import igs_toolbox
+from igs_toolbox.file_utils import read_json_file
 from igs_toolbox.formatChecker.seq_metadata_schema import (
     SeqMetadataKeys,
     ValidationError,
     seq_metadata_schema,
 )
+from igs_toolbox.log_utils import setup_logging
+from igs_toolbox.version_utils import version_callback
 
+if sys.version_info >= (3, 9):
+    from zoneinfo import ZoneInfo
+else:
+    from backports.zoneinfo import ZoneInfo
+NOW = datetime.now(tz=ZoneInfo("Europe/Berlin")).strftime("%Y-%m-%dT%H-%M-%S")  # type: ignore[abstract]
+app = typer.Typer()
 SCHEMA_NAME = "seqMetadata"
 
 
-def validate_species(pathogen: str, species: str) -> bool:
-    """Validate species field."""
+def validate_pathogen_specific_field(pathogen: str, folder: str, value: str) -> bool:
+    """Validate pathogen specific field."""
     # get vocabulary for species
-    answer_set_path = Path(__file__).parent / f"res/species/txt/answerSet{pathogen}.txt"
+    answer_set_path = Path(__file__).parent / f"res/{folder}/txt/valueSet{pathogen}.txt"
     if not answer_set_path.is_file():
         logging.error(f"{answer_set_path} does not point to a file. Aborting.")
         return False
 
-    with Path(answer_set_path).open() as species_file:
-        species_list = [line.strip() for line in species_file]
+    with Path(answer_set_path).open() as file:
+        allowed_values = [line.strip() for line in file]
 
-    if species not in species_list:
-        logging.error(f"{species} is not a valid species for pathogen {pathogen}.")
+    if value not in allowed_values:
         return False
     return True
 
 
+def extract_error_message(schema_error: jsonschema.ValidationError) -> str:
+    if len(schema_error.relative_path) == 0:
+        return schema_error.message
+
+    field_name = schema_error.relative_path.pop()
+    constraint = schema_error.validator
+    raw_message = schema_error.message
+    invalid_value = schema_error.instance
+
+    if isinstance(constraint, str) and constraint == "enum":
+        enum_values = schema_error.validator_value
+        if len(enum_values) > 3:  # noqa: PLR2004
+            values = [repr(v) for v in enum_values[:3]]
+            values.append("...")
+            enum_string = f"[{', '.join(values)}]"
+        else:
+            enum_string = repr(enum_values)
+        raw_message = f"{repr(invalid_value)} is not one of {enum_string}"  # noqa: RUF010
+
+    return f"{field_name}: {repr(invalid_value)} violates constraint {repr(constraint)} ({raw_message})"  # noqa: RUF010
+
+
 def check_seq_metadata(
-    json_data: Dict[Union[SeqMetadataKeys, str], Any],
+    json_data: Dict[str, Any],
     schema: Any = seq_metadata_schema,  # noqa: ANN401
 ) -> None:
     """Validate the sequence metadata."""
-    validator = jsonschema.Draft202012Validator(schema=schema)
-    errors = list(validator.iter_errors(json_data))
-    error_str = []
-    for error in errors:
-        if error.validator == "required":
-            matched_prop = re.search("'(.*)'", error.message)
-            if matched_prop:
-                error_str.append("MISSING_" + matched_prop.group(1))
-            else:
-                error_str.append("MISSING_" + error.message[1 : -len("' is a required property")])
-        else:
-            error_str.append("INVALID_" + error.relative_path[-1])
+    validator = jsonschema.Draft202012Validator(
+        schema=schema,
+        # validating "format" constraints is disabled by default
+        # https://python-jsonschema.readthedocs.io/en/stable/validate/#validating-formats
+        format_checker=jsonschema.Draft202012Validator.FORMAT_CHECKER,
+    )
+    error_messages = [extract_error_message(error) for error in validator.iter_errors(json_data)]
 
-    # some validation.py rules cannot be implemented in jsonschema directly,
-    # thus check them here programmatically
-    if (
-        SeqMetadataKeys.SPECIES in json_data
-        and SeqMetadataKeys.MELDETATBESTAND in json_data
-        and not validate_species(
-            json_data[SeqMetadataKeys.MELDETATBESTAND],
-            json_data[SeqMetadataKeys.SPECIES],
+    # some validation rules cannot be implemented in jsonschema directly, thus check them here programmatically
+    pathogen = json_data.get(SeqMetadataKeys.MELDETATBESTAND)
+    if pathogen:
+        species = json_data.get(SeqMetadataKeys.SPECIES)
+        if species and not validate_pathogen_specific_field(pathogen, SeqMetadataKeys.SPECIES.lower(), species):
+            error_messages.append(f"{repr(species)} is not a valid species for pathogen {pathogen}.")  # noqa: RUF010
+
+        isolation_source = json_data.get(SeqMetadataKeys.ISOLATION_SOURCE)
+        if isolation_source and not validate_pathogen_specific_field(
+            pathogen,
+            SeqMetadataKeys.ISOLATION_SOURCE.lower(),
+            isolation_source,
+        ):
+            error_messages.append(
+                f"{repr(isolation_source)} is not a valid isolation_source for pathogen {pathogen}.",  # noqa: RUF010
+            )
+    else:
+        error_messages.append(
+            "Pathogen is not provided, hence isolation_source and species could not be validated.",
         )
-    ):
-        error_str.append("INVALID_SPECIES")
 
-    if len(error_str) > 0:
-        raise ValidationError("; ".join(error_str))
+    if error_messages:
+        raise ValidationError(error_messages)
 
 
-# Read command line arguments
-def parse(args=None) -> argparse.Namespace:  # noqa: ANN001
-    parser = argparse.ArgumentParser(prog=Path(__file__).name.split(".")[0])
-    parser.add_argument("-i", "--input", required=True, help="Filepath to json file.")
-    parser.add_argument(
-        "-V",
-        "--version",
-        action="version",
-        version=f"%(prog)s {igs_toolbox.__version__}",
-    )
-    return parser.parse_args(args)
-
-
-def main(args=None) -> None:  # noqa: ANN001
-    input_file = Path(parse(args).input)
+@app.command(name="jsonChecker", help="Validate metadata json.")
+def check_json(
+    input_file: Annotated[
+        Path,
+        typer.Option(
+            ...,
+            "--input",
+            "-i",
+            dir_okay=False,
+            file_okay=True,
+            exists=True,
+            help="Path to input json file.",
+        ),
+    ],
+    log_file: Annotated[
+        Path,
+        typer.Option("--log_file", "-l", dir_okay=False, help="Path to log file."),
+    ] = Path(f"./jsonChecker_{NOW}.log"),
+    version: Annotated[  # noqa: ARG001
+        Optional[bool],
+        typer.Option("--version", "-V", callback=version_callback, is_eager=True),
+    ] = None,
+) -> None:
+    setup_logging(log_file=log_file, debug=False)
     # read json file
-    if not input_file.is_file():
-        logging.error(f"{input_file} does not point to a file. Aborting.")
-        sys.exit(1)
-
-    with input_file.open() as jsonfile:
-        try:
-            json_data = json.loads(jsonfile.read())
-        except json.decoder.JSONDecodeError:
-            logging.exception(f"{input_file} is not a valid json file. Aborting.")
-            sys.exit(1)
+    try:
+        json_data = read_json_file(input_file)
+    except FileNotFoundError as e:
+        logging.error(f"{input_file} does not point to a file. Aborting.")  # noqa: TRY400
+        raise typer.Abort(1) from e
+    except json.JSONDecodeError as e:
+        logging.error(f"{input_file} is not a valid json file. Aborting.")  # noqa: TRY400
+        raise typer.Abort from e
 
     # get schema
     try:
         check_seq_metadata(json_data)
-    except ValidationError:
-        logging.exception(f"FAILURE: JSON file does not adhere to the {SCHEMA_NAME} schema.")
-        sys.exit(1)
+    except ValidationError as e:
+        logging.error(f"FAILURE: JSON file does not adhere to the {SCHEMA_NAME} schema.")  # noqa: TRY400
+        for error_message in e.error_messages:
+            logging.error(error_message)  # noqa: TRY400
+        raise typer.Abort from e
 
     logging.info(f"SUCCESS: JSON file adheres to {SCHEMA_NAME} schema.")
     print(f"SUCCESS: JSON file adheres to {SCHEMA_NAME} schema.")  # noqa: T201
 
 
+def main() -> None:
+    """Entry point of CLI tool."""
+    app()
+
+
 if __name__ == "__main__":
     main()
```

## igs_toolbox/formatChecker/seq_metadata_schema.py

```diff
@@ -1,45 +1,58 @@
-from enum import Enum
+from typing import List
 
 
 class ValidationError(Exception):
-    pass
+    def __init__(self, error_messages: List[str]) -> None:
+        super().__init__(
+            "\n".join(
+                [
+                    f"Validation failed ({len(error_messages)} issues)",
+                    *[f"- {message}" for message in error_messages],
+                ],
+            ),
+        )
+        self.error_messages = error_messages
 
 
-class SeqMetadataKeys(str, Enum):
+class SeqMetadataKeys:
     MELDETATBESTAND = "MELDETATBESTAND"
     SPECIES = "SPECIES"
     LAB_SEQUENCE_ID = "LAB_SEQUENCE_ID"
     DEMIS_NOTIFICATION_ID = "DEMIS_NOTIFICATION_ID"
     STATUS = "STATUS"
+    VERSION = "VERSION"
     IGS_ID = "IGS_ID"
     DATE_OF_RECEIVING = "DATE_OF_RECEIVING"
     DATE_OF_SAMPLING = "DATE_OF_SAMPLING"
     DATE_OF_SEQUENCING = "DATE_OF_SEQUENCING"
     DATE_OF_SUBMISSION = "DATE_OF_SUBMISSION"
+    DATE_OF_DELETION = "DATE_OF_DELETION"
     SEQUENCING_INSTRUMENT = "SEQUENCING_INSTRUMENT"
     SEQUENCING_PLATFORM = "SEQUENCING_PLATFORM"
     ADAPTER = "ADAPTER"
     SEQUENCING_STRATEGY = "SEQUENCING_STRATEGY"
     ISOLATION_SOURCE = "ISOLATION_SOURCE"
     HOST = "HOST"
     SEQUENCING_REASON = "SEQUENCING_REASON"
     GEOGRAPHIC_LOCATION = "GEOGRAPHIC_LOCATION"
     ISOLATE = "ISOLATE"
     AUTHOR = "AUTHOR"
     NAME_AMP_PROTOCOL = "NAME_AMP_PROTOCOL"
     PRIMER_SCHEME = "PRIMER_SCHEME"
+    METADATA_QC = "METADATA_QC"
+    METADATA_QC_REASON = "METADATA_QC_REASON"
     PRIME_DIAGNOSTIC_LAB_DEMIS_LAB_ID = "PRIME_DIAGNOSTIC_LAB.DEMIS_LAB_ID"
     PRIME_DIAGNOSTIC_LAB_NAME = "PRIME_DIAGNOSTIC_LAB.NAME"
-    PRIME_DIAGNOSTIC_LAB_ADDRESS = "PRIME_DIAGNOSTIC_LAB.STREET"
+    PRIME_DIAGNOSTIC_LAB_ADDRESS = "PRIME_DIAGNOSTIC_LAB.ADDRESS"
     PRIME_DIAGNOSTIC_LAB_POSTAL_CODE = "PRIME_DIAGNOSTIC_LAB.POSTAL_CODE"
     PRIME_DIAGNOSTIC_LAB_FEDERAL_STATE = "PRIME_DIAGNOSTIC_LAB.FEDERAL_STATE"
     SEQUENCING_LAB_DEMIS_LAB_ID = "SEQUENCING_LAB.DEMIS_LAB_ID"
     SEQUENCING_LAB_NAME = "SEQUENCING_LAB.NAME"
-    SEQUENCING_LAB_ADDRESS = "SEQUENCING_LAB.STREET"
+    SEQUENCING_LAB_ADDRESS = "SEQUENCING_LAB.ADDRESS"
     SEQUENCING_LAB_POSTAL_CODE = "SEQUENCING_LAB.POSTAL_CODE"
     SEQUENCING_LAB_FEDERAL_STATE = "SEQUENCING_LAB.FEDERAL_STATE"
     UPLOADS = "Uploads"
     REPOSITORY_NAME = "REPOSITORY_NAME"
     REPOSITORY_LINK = "REPOSITORY_LINK"
     REPOSITORY_ID = "REPOSITORY_ID"
     UPLOAD_DATE = "UPLOAD_DATE"
@@ -48,14 +61,15 @@
     FILES = "Files"
     FILE_NAME = "FILE_NAME"
     FILE_SHA256SUM = "FILE_SHA256SUM"
 
 
 seq_metadata_schema = {
     "type": "object",
+    "additionalProperties": False,
     "properties": {
         SeqMetadataKeys.MELDETATBESTAND: {
             "enum": [
                 "EHCP",
                 "LISP",
                 "SALP",
                 "STYP",
@@ -74,29 +88,45 @@
                 "HEVP",
                 "HAVP",
                 "LEGP",
                 "SPNP",
                 "WNVP",
             ],
         },
-        SeqMetadataKeys.SPECIES: {"type": "string", "pattern": "[A-z0-9\\-_]*"},
+        SeqMetadataKeys.SPECIES: {
+            "type": "string",
+            "minLength": 1,
+        },
         SeqMetadataKeys.LAB_SEQUENCE_ID: {
             "type": "string",
-            "pattern": "[A-z,0-9,-,_]*",
+            "pattern": "^[A-Za-z0-9-_]+$",
+        },
+        SeqMetadataKeys.DEMIS_NOTIFICATION_ID: {
+            "type": "string",
+            "minLength": 1,
         },
-        SeqMetadataKeys.DEMIS_NOTIFICATION_ID: {"type": "string"},
-        SeqMetadataKeys.STATUS: {"type": "string"},
+        SeqMetadataKeys.STATUS: {
+            "type": "string",
+            "minLength": 1,
+        },
+        SeqMetadataKeys.VERSION: {"$ref": "#/$defs/int"},
         SeqMetadataKeys.IGS_ID: {
             "type": "string",
-            "pattern": "(IMS|IGS)-[0-9]{5}-[A-Z]{4}-[A-z0-9\\-_]+",
+            "pattern": (
+                "^(IMS|IGS)-[0-9]{5}-[A-Z]{3}P-"
+                "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"
+                "|"
+                "^(IMS|IGS)-[0-9]{5}-[A-Z]{3}P-[0-9]{2,6}$"  # for migration of old sequences
+            ),
         },
         SeqMetadataKeys.DATE_OF_RECEIVING: {"$ref": "#/$defs/date"},
         SeqMetadataKeys.DATE_OF_SAMPLING: {"$ref": "#/$defs/date"},
         SeqMetadataKeys.DATE_OF_SEQUENCING: {"$ref": "#/$defs/date"},
         SeqMetadataKeys.DATE_OF_SUBMISSION: {"$ref": "#/$defs/date"},
+        SeqMetadataKeys.DATE_OF_DELETION: {"$ref": "#/$defs/date"},
         SeqMetadataKeys.SEQUENCING_INSTRUMENT: {
             "enum": [
                 "454_GS",
                 "454_GS_20",
                 "454_GS_FLX",
                 "454_GS_FLX+",
                 "454_GS_FLX_Titanium",
@@ -171,15 +201,15 @@
                 "BGISEQ",
                 "DNBSEQ",
                 "OTHER",
             ],
         },
         SeqMetadataKeys.ADAPTER: {
             "type": "string",
-            "pattern": "[a-zA-Z0-9_\\- ,]+",
+            "pattern": "^[\\+a-zA-Z0-9_ ,-]+$",
         },
         SeqMetadataKeys.SEQUENCING_STRATEGY: {
             "enum": [
                 "WGS",
                 "WGA",
                 "WXS",
                 "RNA-Seq",
@@ -214,568 +244,78 @@
                 "Synthetic-Long-Read",
                 "Targeted-Capture",
                 "Tethered Chromatin Conformation Capture",
                 "OTHER",
             ],
         },
         SeqMetadataKeys.ISOLATION_SOURCE: {
-            "enum": [
-                "Air sample",
-                "Ear sample",
-                "Pus sample",
-                "Bile sample",
-                "Tick sample",
-                "Milk sample",
-                "Hair sample",
-                "Vein sample",
-                "Lens sample",
-                "Tear sample",
-                "Skin sample",
-                "Nail sample",
-                "Water sample",
-                "Serum sample",
-                "Joint sample",
-                "Smear sample",
-                "Feces sample",
-                "Sperm sample",
-                "Stone sample",
-                "Vomit sample",
-                "Lymph sample",
-                "Mucus sample",
-                "Gonad sample",
-                "Larva sample",
-                "Cream sample",
-                "Bursa sample",
-                "Sebum sample",
-                "Fluid sample",
-                "Blood sample",
-                "Urine sample",
-                "Louse sample",
-                "Lesion sample",
-                "Tendon sample",
-                "Fascia sample",
-                "Living sample",
-                "Maggot sample",
-                "Plasma sample",
-                "Lochia sample",
-                "Breast sample",
-                "Artery sample",
-                "Biopsy sample",
-                "Insect sample",
-                "Liquor sample",
-                "Sputum sample",
-                "Neuroma sample",
-                "Genetic sample",
-                "Genital sample",
-                "Exudate sample",
-                "Cytology sample",
-                "Synovial sample",
-                "Ear swab sample",
-                "Thrombus sample",
-                "Parasite sample",
-                "Soy milk sample",
-                "Platelet sample",
-                "Prostate sample",
-                "Synovium sample",
-                "Hematoma sample",
-                "Ligament sample",
-                "Appendix sample",
-                "Effusion sample",
-                "Helminth sample",
-                "Meconium sample",
-                "Cartilage sample",
-                "Eye fluid sample",
-                "Pituitary sample",
-                "Colostrum sample",
-                "Hydrocele sample",
-                "Colectomy sample",
-                "Gallstone sample",
-                "Endocrine sample",
-                "Skin cyst sample",
-                "Hair bulb sample",
-                "Dialysate sample",
-                "Hot stool sample",
-                "Ice-cream sample",
-                "Foodstuff sample",
-                "Wash fluid sample",
-                "Urological sample",
-                "Cyst fluid sample",
-                "Skin scale sample",
-                "Whole milk sample",
-                "Cord blood sample",
-                "Body fluid sample",
-                "Epididymis sample",
-                "Anal polyp sample",
-                "Lip biopsy sample",
-                "Blood film sample",
-                "Polar body sample",
-                "Anal smear sample",
-                "Cystectomy sample",
-                "Spot urine sample",
-                "Transudate sample",
-                "Lymph node sample",
-                "Cow's milk sample",
-                "Ear tissue sample",
-                "Eye tissue sample",
-                "Fibroblast sample",
-                "Soft stool sample",
-                "Paint chip sample",
-                "Blood clot sample",
-                "Wound fluid sample",
-                "Lung biopsy sample",
-                "Skin lesion sample",
-                "Nasal polyp sample",
-                "Omentectomy sample",
-                "Breast duct sample",
-                "Bursa fluid sample",
-                "Sinus fluid sample",
-                "Whole blood sample",
-                "Goat's milk sample",
-                "Labia ulcer sample",
-                "Gastrectomy sample",
-                "Bone biopsy sample",
-                "Parathyroid sample",
-                "Oral cavity sample",
-                "Colon strip sample",
-                "Anal biopsy sample",
-                "Fecal fluid sample",
-                "Edema fluid sample",
-                "Oral saliva sample",
-                "Labia smear sample",
-                "Dental cyst sample",
-                "Skin biopsy sample",
-                "Soft tissue sample",
-                "Oral tissue sample",
-                "Nephrectomy sample",
-                "Nasal fluid sample",
-                "Bone tissue sample",
-                "Rectal snip sample",
-                "Renal stone sample",
-                "Heart valve sample",
-                "Respiratory sample",
-                "Anal tissue sample",
-                "Bone marrow sample",
-                "Breast cyst sample",
-                "Nerve tissue sample",
-                "Orchidectomy sample",
-                "Mouth biopsy sample",
-                "Renal biopsy sample",
-                "Penis tissue sample",
-                "Formula milk sample",
-                "Excised bone sample",
-                "Cecum biopsy sample",
-                "Reticulocyte sample",
-                "Venous blood sample",
-                "Liver tissue sample",
-                "Colon tissue sample",
-                "Bursa tissue sample",
-                "Bursa biopsy sample",
-                "Joint biopsy sample",
-                "Nerve biopsy sample",
-                "Vas deferens sample",
-                "Excised lung sample",
-                "Laryngectomy sample",
-                "Liver biopsy sample",
-                "Ileal biopsy sample",
-                "Tumor tissue sample",
-                "Buccal smear sample",
-                "Hemofiltrate sample",
-                "Male genital sample",
-                "Breast fluid sample",
-                "Thyroid lobe sample",
-                "Nasal biopsy sample",
-                "Rectal polyp sample",
-                "Ovarian cyst sample",
-                "Hysterectomy sample",
-                "Tissue fluid sample",
-                "Skimmed milk sample",
-                "Sheep's milk sample",
-                "Aorta tissue sample",
-                "Kidney tissue sample",
-                "Lipoma biopsy sample",
-                "Excised bursa sample",
-                "Fascia biopsy sample",
-                "Testes biopsy sample",
-                "Pleura biopsy sample",
-                "Ureter biopsy sample",
-                "Muscle biopsy sample",
-                "Testes tissue sample",
-                "Uterus tissue sample",
-                "Cervix tissue sample",
-                "Aqueous humor sample",
-                "Chylous fluid sample",
-                "Empyema fluid sample",
-                "Jejunal fluid sample",
-                "Biliary stone sample",
-                "Bladder stone sample",
-                "Amniotic cell sample",
-                "Clotted blood sample",
-                "Blister fluid sample",
-                "Thyroid gland sample",
-                "Head of femur sample",
-                "Thyroid fluid sample",
-                "Larynx biopsy sample",
-                "Dental tissue sample",
-                "Tonsil biopsy sample",
-                "Rectal tissue sample",
-                "Ascitic fluid sample",
-                "Embryo biopsy sample",
-                "Fistula fluid sample",
-                "Genitourinary sample",
-                "Tendon biopsy sample",
-                "Adrenal gland sample",
-                "Tongue biopsy sample",
-                "Tonsillectomy sample",
-                "Gastric ulcer sample",
-                "Colonic polyp sample",
-                "Rectal biopsy sample",
-                "Ovarian fluid sample",
-                "Parotidectomy sample",
-                "Rectal scrape sample",
-                "Gastric fluid sample",
-                "Thyroidectomy sample",
-                "Maternal milk sample",
-                "Breast biopsy sample",
-                "Desmoid tumor sample",
-                "Treated water sample",
-                "24 hour urine sample",
-                "Red blood cell sample",
-                "Adrenal biopsy sample",
-                "Gastric biopsy sample",
-                "Colonic biopsy sample",
-                "Synovial fluid sample",
-                "Adrenal tissue sample",
-                "Cervical smear sample",
-                "Adipose tissue sample",
-                "Oral secretion sample",
-                "Dialysis fluid sample",
-                "Duodenal fluid sample",
-                "Gastric tissue sample",
-                "Mastoid biopsy sample",
-                "Omentum biopsy sample",
-                "Jejunal biopsy sample",
-                "Anal resection sample",
-                "Vaginal biopsy sample",
-                "Omentum tissue sample",
-                "Ureteric stone sample",
-                "Bone and joint sample",
-                "Gingiva biopsy sample",
-                "Body substance sample",
-                "Female genital sample",
-                "Excised tendon sample",
-                "Excised fascia sample",
-                "Excised rectum sample",
-                "Bladder biopsy sample",
-                "Urethra biopsy sample",
-                "Cervical polyp sample",
-                "Pancreatectomy sample",
-                "Vitreous humor sample",
-                "Drainage fluid sample",
-                "Cardiovascular sample",
-                "Dermatological sample",
-                "Ovarian tissue sample",
-                "Vaginal tissue sample",
-                "Arterial blood sample",
-                "Ovarian biopsy sample",
-                "Duodenal ulcer sample",
-                "Capillary blood sample",
-                "Heart sac fluid sample",
-                "Skin curettings sample",
-                "Musculoskeletal sample",
-                "Placenta biopsy sample",
-                "Hydrocele fluid sample",
-                "Prostatic fluid sample",
-                "Pancreas tissue sample",
-                "Synovium biopsy sample",
-                "Prostate biopsy sample",
-                "Cholecystectomy sample",
-                "Pancreas biopsy sample",
-                "Cervical biopsy sample",
-                "Temporal artery sample",
-                "Total colectomy sample",
-                "Foreskin biopsy sample",
-                "Chorionic villi sample",
-                "Epidural lesion sample",
-                "Ligament biopsy sample",
-                "Tracheal biopsy sample",
-                "Duodenal biopsy sample",
-                "Ileal resection sample",
-                "Rib bone marrow sample",
-                "Colostomy fluid sample",
-                "Bronchial fluid sample",
-                "Ileostomy fluid sample",
-                "Skin fibroblast sample",
-                "Acidified serum sample",
-                "Prostate tissue sample",
-                "Untreated water sample",
-                "Abdominal fluid sample",
-                "Sputum - coughed sample",
-                "Stomach aspirate sample",
-                "First void urine sample",
-                "Bronchial biopsy sample",
-                "Urological fluid sample",
-                "Placental tissue sample",
-                "Middle ear fluid sample",
-                "Fecal concretion sample",
-                "Mid-stream urine sample",
-                "White blood cell sample",
-                "Kidney brushings sample",
-                "Pancreatic fluid sample",
-                "Excised ligament sample",
-                "Bile duct biopsy sample",
-                "Subretinal fluid sample",
-                "Corneal scraping sample",
-                "Pancreatic stone sample",
-                "Bone marrow clot sample",
-                "Intraocular lens sample",
-                "Brain cyst fluid sample",
-                "Spleen resection sample",
-                "Cartilage biopsy sample",
-                "Articular tissue sample",
-                "Peritoneal fluid sample",
-                "Gastrointestinal sample",
-                "Total gastrectomy sample",
-                "Respiratory fluid sample",
-                "Penis skin biopsy sample",
-                "Lymph node tissue sample",
-                "Middle ear biopsy sample",
-                "Peritoneal biopsy sample",
-                "Esophageal biopsy sample",
-                "Gastric brushings sample",
-                "Endometrial polyp sample",
-                "Skin punch biopsy sample",
-                "Semi-skimmed milk sample",
-                "Esophageal tissue sample",
-                "Excised cartilage sample",
-                "Resected prostate sample",
-                "Vocal cord biopsy sample",
-                "Pharyngeal biopsy sample",
-                "Lumpectomy breast sample",
-                "Vaginal secretion sample",
-                "Upper respiratory sample",
-                "Extradural lesion sample",
-                "Lymph node biopsy sample",
-                "Lung direct smear sample",
-                "Partial colectomy sample",
-                "Total nephrectomy sample",
-                "Breast cyst fluid sample",
-                "Postmortem tissue sample",
-                "Lower respiratory sample",
-                "Surgical excision sample",
-                "Urological tissue sample",
-                "Second void urine sample",
-                "Endotracheal fluid sample",
-                "Vulval skin biopsy sample",
-                "Parathyroid biopsy sample",
-                "Liver direct smear sample",
-                "Small bowel tissue sample",
-                "Renal pelvis fluid sample",
-                "Excised mesorectum sample",
-                "Endometrial biopsy sample",
-                "Native heart valve sample",
-                "Conjunctival smear sample",
-                "Nasal lavage fluid sample",
-                "Endocardium biopsy sample",
-                "Soft tissue biopsy sample",
-                "Thyroid cyst fluid sample",
-                "Parathyroid tissue sample",
-                "Bladder curettings sample",
-                "Large bowel tissue sample",
-                "Ureteric brushings sample",
-                "Endocervical smear sample",
-                "Papanicolaou smear sample",
-                "Duodenal brushings sample",
-                "Endocervical polyp sample",
-                "Peritoneal dialysis sample",
-                "Amniocentesis fluid sample",
-                "Partial nephrectomy sample",
-                "Sternal bone marrow sample",
-                "Cerebrospinal fluid sample",
-                "Transtracheal fluid sample",
-                "Early morning urine sample",
-                "Uterine cervix cyst sample",
-                "Thoracentesis fluid sample",
-                "Right hemicolectomy sample",
-                "Excised parathyroid sample",
-                "Bronchial brushings sample",
-                "Vaginal vault smear sample",
-                "Breast direct smear sample",
-                "Suprapubic aspirate sample",
-                "Total parotidectomy sample",
-                "Pancreas cyst fluid sample",
-                "Cerebellar neoplasm sample",
-                "Core sample of tissue block",
-                "Thyroid direct smear sample",
-                "Esophageal brushings sample",
-                "Sigmoid colon biopsy sample",
-                "Biliary tract tissue sample",
-                "Skin excision biopsy sample",
-                "Branchial cleft cyst sample",
-                "Minor salivary gland sample",
-                "Bone marrow trephine sample",
-                "Nasal sinus contents sample",
-                "Vaginal vault biopsy sample",
-                "Skin incision biopsy sample",
-                "Segmentectomy breast sample",
-                "Red blood cell fluid sample",
-                "Parotid gland saliva sample",
-                "Bone marrow fragment sample",
-                "Female genital fluid sample",
-                "Breast tru-cut biopsy sample",
-                "Frozen section breast sample",
-                "Excised adrenal gland sample",
-                "Nasopharyngeal biopsy sample",
-                "Gastric antral biopsy sample",
-                "Gastric fundus biopsy sample",
-                "Sigmoid colonic polyp sample",
-                "Neck block dissection sample",
-                "Vaginal washout fluid sample",
-                "Machine hemodialysate sample",
-                "Brain neoplasm tissue sample",
-                "Frozen tissue section sample",
-                "Small bowel resection sample",
-                "Endometrial resection sample",
-                "Deoxyribonucleic acid sample",
-                "Contact lens solution sample",
-                "Female genital tissue sample",
-                "Fallopian tube biopsy sample",
-                "Umbilical cord tissue sample",
-                "Endometrial curettings sample",
-                "Urinary bladder tissue sample",
-                "Frozen section bladder sample",
-                "Cervical loop excision sample",
-                "Ascending colon biopsy sample",
-                "Upper respiratory swab sample",
-                "Excised salivary gland sample",
-                "Gastrointestinal fluid sample",
-                "Retroperitoneal tissue sample",
-                "Ileocecal valve biopsy sample",
-                "Gingivocrevicular fluid sample",
-                "Pancreaticoduodenectomy sample",
-                "Gastric lavage aspirate sample",
-                "Renal transplant biopsy sample",
-                "Upper respiratory fluid sample",
-                "Prostatic massage fluid sample",
-                "Oral mucosal transudate sample",
-                "Prostate tru-cut biopsy sample",
-                "Lower respiratory fluid sample",
-                "Sigmoid colon brushings sample",
-                "Breast nipple discharge sample",
-                "Iliac crest bone marrow sample",
-                "Tissue impression smear sample",
-                "Ampulla of Vater biopsy sample",
-                "Gastrointestinal tissue sample",
-                "Descending colon biopsy sample",
-                "Thyroid neoplasm tissue sample",
-                "Terminal ileum resection sample",
-                "Upper respiratory tissue sample",
-                "Lower respiratory tissue sample",
-                "Excised sublingual gland sample",
-                "Superficial parotidectomy sample",
-                "Resected articular tissue sample",
-                "Frozen section lymph node sample",
-                "Placental membrane tissue sample",
-                "Lung fine needle aspirate sample",
-                "Cerebrospinal fluid smear sample",
-                "Left colic flexure biopsy sample",
-                "Rectosigmoid colon biopsy sample",
-                "Intravenous infusion fluid sample",
-                "Right colic flexure biopsy sample",
-                "Common bile duct brushings sample",
-                "Transbronchial lung biopsy sample",
-                "Excised submandibular gland sample",
-                "Transthoracic needle biopsy sample",
-                "Breast fine needle aspirate sample",
-                "Thyroid fine needle aspirate sample",
-                "Anterior resection of rectum sample",
-                "Bronchoalveolar lavage fluid sample",
-                "Products of conception tissue sample",
-                "Excised breast ectopic tissue sample",
-                "Intravenous lipid infusion fluid sample",
-                "Central nervous system cyst fluid sample",
-                "Donor-organ preservation solution sample",
-                "Cardiopulmonary bypass pump fluid sample",
-                "Colorectal junction neoplasm tissue sample",
-                "Vertebral spinous process bone marrow sample",
-                "Soft tissue lesion fine needle aspirate sample",
-                "Frozen section skin sample obtained by excision",
-                "Hysterectomy and left salpingo-oophorectomy sample",
-                "Excised uterus, fallopian tubes and ovaries sample",
-                "Hydrocele sample obtained by fine needle aspiration",
-                "Hysterectomy and right salpingo-oophorectomy sample",
-                "Tissue sample from aneurysm obtained by surgical resection",
-                "Nasopharyngeal swab",
-                "Swab from nasal sinus",
-                "Anterior nares swab",
-                "Oropharyngeal aspirate",
-                "Nasopharyngeal aspirate",
-                "Specimen from trachea obtained by aspiration",
-                "Pleural fluid specimen",
-                "Nasopharyngeal washings",
-                "Plasma specimen or serum specimen or whole blood specimen",
-                "Swab of internal nose",
-                "Throat swab",
-                "Specimen unsatisfactory for evaluation (finding)",
-                "Screening - procedure intent",
-            ],
+            "type": "string",
+            "minLength": 1,
         },
         SeqMetadataKeys.HOST: {"enum": ["Homo sapiens"]},
         SeqMetadataKeys.SEQUENCING_REASON: {"enum": ["random", "requested", "other", "clinical"]},
         SeqMetadataKeys.GEOGRAPHIC_LOCATION: {
             "type": "string",
-            "pattern": "[0-9]{3}$",
+            "pattern": "^[0-9]{3}$",
         },
         SeqMetadataKeys.ISOLATE: {
             "type": "string",
-            "pattern": "[a-zA-Z0-9_\\- ,]*",
+            "minLength": 1,
         },
         SeqMetadataKeys.AUTHOR: {
             "type": "string",
-            "pattern": "[a-zA-Z0-9_\\- ,]*",
+            "minLength": 1,
         },
         SeqMetadataKeys.NAME_AMP_PROTOCOL: {
             "type": "string",
-            "pattern": "[a-zA-Z0-9_\\- ,]*",
+            "pattern": "^[a-zA-Z0-9_ ,-]+$",
         },
         SeqMetadataKeys.PRIME_DIAGNOSTIC_LAB_DEMIS_LAB_ID: {
             "type": "string",
-            "pattern": "^$|DEMIS-[0-9]{5}",
+            "pattern": "^DEMIS-[0-9]{5}",
         },
         SeqMetadataKeys.PRIME_DIAGNOSTIC_LAB_NAME: {
             "type": "string",
-            "pattern": "[a-zA-Z0-9_\\- ,]*",
+            "pattern": "^[a-zA-Z0-9_ ,-]+$",
         },
         SeqMetadataKeys.PRIME_DIAGNOSTIC_LAB_ADDRESS: {
             "type": "string",
-            "pattern": "[a-zA-Z0-9_\\- ,]*",
+            "minLength": 1,
         },
         SeqMetadataKeys.PRIME_DIAGNOSTIC_LAB_POSTAL_CODE: {
             "type": "string",
-            "pattern": "[0-9]{5}|^$",
+            "pattern": "^[0-9]{5}$",
         },
         SeqMetadataKeys.PRIME_DIAGNOSTIC_LAB_FEDERAL_STATE: {"$ref": "#/$defs/federalStates"},
         SeqMetadataKeys.PRIMER_SCHEME: {
             "type": "string",
-            "pattern": "[a-zA-Z0-9_\\- ,]*",
+            "pattern": "^[a-zA-Z0-9_ ,-\\.]+$",
+        },
+        SeqMetadataKeys.METADATA_QC: {
+            "$ref": "#/$defs/bool",
+        },
+        SeqMetadataKeys.METADATA_QC_REASON: {
+            "type": "string",
+            "pattern": "^[a-zA-Z0-9_ ,-]+$",
         },
         SeqMetadataKeys.SEQUENCING_LAB_DEMIS_LAB_ID: {
             "type": "string",
-            "pattern": "^$|DEMIS-[0-9]{5}",
+            "pattern": "^DEMIS-[0-9]{5}$",
         },
         SeqMetadataKeys.SEQUENCING_LAB_NAME: {
             "type": "string",
-            "pattern": "[a-zA-Z0-9_\\- ,]*",
+            "pattern": "^[a-zA-Z0-9_ ,-]+$",
         },
         SeqMetadataKeys.SEQUENCING_LAB_ADDRESS: {
             "type": "string",
-            "pattern": "[a-zA-Z0-9_\\- ,]*",
+            "minLength": 1,
         },
         SeqMetadataKeys.SEQUENCING_LAB_POSTAL_CODE: {
             "type": "string",
-            "pattern": "[0-9]{5}$",
+            "pattern": "^[0-9]{5}$",
         },
         SeqMetadataKeys.SEQUENCING_LAB_FEDERAL_STATE: {"$ref": "#/$defs/federalStates"},
         SeqMetadataKeys.UPLOADS: {
             "type": "array",
             "items": {
                 "type": "object",
                 "properties": {
@@ -789,44 +329,48 @@
                 "additionalProperties": False,
                 "required": [
                     SeqMetadataKeys.REPOSITORY_NAME,
                     SeqMetadataKeys.REPOSITORY_ID,
                     SeqMetadataKeys.UPLOAD_STATUS,
                 ],
             },
-            "minItems": 1,
         },
         SeqMetadataKeys.FILES: {
             "type": "array",
             "items": {
                 "type": "object",
                 "properties": {
                     SeqMetadataKeys.FILE_NAME: {
                         "type": "string",
-                        "pattern": "^[a-zA-Z0-9_\\-]+.(fasta|fastq.gz|fq.gz|.fq)",
+                        "pattern": "^[a-zA-Z0-9_-]+\\.(fasta|fa|fastq|fq)(\\.gz)?$",
                     },
                     SeqMetadataKeys.FILE_SHA256SUM: {
                         "type": "string",
-                        "pattern": "[A-Fa-f0-9]{64}",
+                        "pattern": "^[A-Fa-f0-9]{64}$",
                     },
                 },
                 "required": ["FILE_NAME", "FILE_SHA256SUM"],
                 "additionalProperties": False,
             },
             "minItems": 1,
             "uniqueItems": True,
         },
     },
     "$defs": {
+        "bool": {
+            "type": "string",
+            "pattern": "^(true)|(false)$",
+        },
+        "int": {
+            "type": "string",
+            "pattern": "^-?[0-9]+$",
+        },
         "date": {
             "type": "string",
             "format": "date",
-            "pattern": "(((19|20)([2468][048]|[13579][26]|0[48])|2000)[/-]02[/-]29|((19|20)[0-9]{2}[/-](0[469]|11)[/-](0[1-9]|[12][0-9]|30)|(19|20)[0-9]{2}[/-](0[13578]|1[02])[/-](0[1-9]|[12][0-9]|3[01])|(19|20)[0-9]{2}[/-]02[/-](0[1-9]|1[0-9]|2[0-8]{1})))",  # noqa: E501
-            "minLength": 10,
-            "maxLength": 10,
         },
         "federalStates": {
             "enum": [
                 "Baden-Wrttemberg",
                 "Bayern",
                 "Berlin",
                 "Brandenburg",
```

## Comparing `igs_toolbox/formatChecker/res/species/txt/answerSetACBP.txt` & `igs_toolbox/formatChecker/res/species/txt/valueSetACBP.txt`

 * *Files identical despite different names*

## Comparing `igs_toolbox/formatChecker/res/species/txt/answerSetCVDP.txt` & `igs_toolbox/formatChecker/res/species/txt/valueSetCVDP.txt`

 * *Files identical despite different names*

## Comparing `igs_toolbox/formatChecker/res/species/txt/answerSetEBCP.txt` & `igs_toolbox/formatChecker/res/species/txt/valueSetEBCP.txt`

 * *Files identical despite different names*

## Comparing `igs_toolbox/formatChecker/res/species/txt/answerSetEHCP.txt` & `igs_toolbox/formatChecker/res/species/txt/valueSetEHCP.txt`

 * *Files identical despite different names*

## Comparing `igs_toolbox/formatChecker/res/species/txt/answerSetLEGP.txt` & `igs_toolbox/formatChecker/res/species/txt/valueSetLEGP.txt`

 * *Files identical despite different names*

## Comparing `igs_toolbox/formatChecker/res/species/txt/answerSetLISP.txt` & `igs_toolbox/formatChecker/res/species/txt/valueSetLISP.txt`

 * *Files identical despite different names*

## Comparing `igs_toolbox/formatChecker/res/species/txt/answerSetMSVP.txt` & `igs_toolbox/formatChecker/res/species/txt/valueSetMSVP.txt`

 * *Files identical despite different names*

## Comparing `igs_toolbox/formatChecker/res/species/txt/answerSetMYTP.txt` & `igs_toolbox/formatChecker/res/species/txt/valueSetMYTP.txt`

 * *Files identical despite different names*

## Comparing `igs_toolbox/formatChecker/res/species/txt/answerSetNEIP.txt` & `igs_toolbox/formatChecker/res/species/txt/valueSetNEIP.txt`

 * *Files identical despite different names*

## Comparing `igs_toolbox/formatChecker/res/species/txt/answerSetSALP.txt` & `igs_toolbox/formatChecker/res/species/txt/valueSetSALP.txt`

 * *Files identical despite different names*

## Comparing `igs_toolbox/formatChecker/res/species/txt/answerSetSPNP.txt` & `igs_toolbox/formatChecker/res/species/txt/valueSetSPNP.txt`

 * *Files identical despite different names*

## Comparing `igs_toolbox-1.5.2.dist-info/LICENSE` & `igs_toolbox-1.5.7.dist-info/LICENSE`

 * *Files identical despite different names*

## Comparing `igs_toolbox-1.5.2.dist-info/METADATA` & `igs_toolbox-1.5.7.dist-info/METADATA`

 * *Files 3% similar despite different names*

```diff
@@ -1,10 +1,10 @@
 Metadata-Version: 2.1
 Name: igs-toolbox
-Version: 1.5.2
+Version: 1.5.7
 Summary: A toolbox to check whether files follow a predefined schema.
 Author-email: IGS Developers <IGS-Developers@rki.de>
 License: The MIT License (MIT)
         
         Copyright (c) 2022 NA
         
         Permission is hereby granted, free of charge, to any person obtaining a copy
@@ -38,14 +38,16 @@
 Requires-Dist: backports.zoneinfo ; python_version < "3.9"
 Provides-Extra: dev
 Requires-Dist: black ; extra == 'dev'
 Requires-Dist: mypy ; extra == 'dev'
 Requires-Dist: ruff ; extra == 'dev'
 Requires-Dist: pytest ; extra == 'dev'
 Requires-Dist: pytest-cov ; extra == 'dev'
+Requires-Dist: types-jsonschema ; extra == 'dev'
+Requires-Dist: pandas-stubs ; extra == 'dev'
 
 # igs-toolbox
 
 ## Description
 This Python package contains tools to check whether files follow predefined schemas of the IGS project.
 Currently, only a JSON validator is implemented, but the idea is that it will be extended to more applications related to the IGS project in the future.
```

